lib/spack/spack/installer.py

```diff
@@ -2494,6 +2494,15 @@ def build_process(pkg: "spack.package_base.PackageBase", install_args: dict) ->
 
 def deprecate(spec: "spack.spec.Spec", deprecator: "spack.spec.Spec", link_fn) -> None:
     """Deprecate this package in favor of deprecator spec"""
+    # Here we assume we don't deprecate across different stores, and that same hash
+    # means same binary artifacts
+    if spec.dag_hash() == deprecator.dag_hash():
```

Comment:
shouldn't this just be `spec == deprecator`?

Reply:
Hmm, we could use that, but does it really matter? `==` would use `Spec._cmp_iter` and first compare `name` etc. and then compare the `Spec.process_hash()`